crypto/sparse_array/README

```diff
@@ -0,0 +1,55 @@
+This directory contains an implementation of a sparse array that attempts
+to be both space and time efficient.
+
+The sparse array is represented using a tree structure.  Each node in the
+tree contains a block of pointers to either the user supplied leaf values or
+to another node.
+
+There are a number of parameters used to define the block size:
+
+    SA_BLOCK_BITS           Specifies the number of bits covered by each block
+    SA_BLOCK_MAX            Specifies the number of pointers in each block
+    SA_BLOCK_MASK           Specifies a bit mask to perform modulo block size
+    SA_BLOCK_MAX_LEVELS     Indicates the maximum possible height of the tree
+
+These constants are inter-related:
+    SA_BLOCK_MAX        = 2 ^ SA_BLOCK_BITS
+    SA_BLOCK_MASK       = SA_BLOCK_MAX - 1
+    SA_BLOCK_MAX_LEVELS = number of bits in size_t divided by SA_BLOCK_BITS
+
+As a space and performance optimisation, the height of the tree is usually
+less than the maximum possible height.  Only sufficient height is allocated
+to accommodate the largest index added to the data structure.
+
+The largest index used to add a value to the array determines the tree
+height:
+
+        +----------------------+---------------------+
+        | Largest Added Index  |   Height of Tree    |
+        +----------------------+---------------------+
+        | SA_BLOCK_MAX     - 1 |          1          |
+        | SA_BLOCK_MAX ^ 2 - 1 |          2          |
+        | SA_BLOCK_MAX ^ 3 - 1 |          3          |
+        | ...                  |          ...        |
+        | size_t max           | SA_BLOCK_MAX_LEVELS |
+        +----------------------+---------------------+
+
+The tree height is dynamically increased as needed based on additions.
+
+The nodes themselves are allocated in a sparse manner.  Only nodes which
+exist along a path from the root of the tree to an added leaf will be
+allocated.  Again, the complexity is hidden and nodes are allocated on an
+as needed basis.
+
+Values can be removed from the sparse array by setting their index
+position to NULL.  The data structure does not attempt to reclaim nodes
+or reduce the height of the tree on removal.
+
+Accesses to elements in the sparse array take O(log n) time where n is
+the largest element.  The base of the logarithm is SA_BLOCK_MAX, so for
+moderately small indices, single level (constant time) access is achievable.
+Space usage is O(m) where m is the number of elements in the array.
+
+Note: sparse arrays only include pointers to types.  This SPARSE_ARRAY_OF(char)
```

Comment:
Is the second sentence supposed to start with Thus?

Reply:
Yes, changed.  Thanks once again.

Making the node size variable isn't difficult.  Replace the initial macros in `sparse_array.c` with fields in the base structure (i.e.  `SA_BLOCK_BITS`, `SA_BLOCK_MAX`, `SA_BLOCK_MASK`, and `SA_BLOCK_MAX_LEVELS`) and add a call to set the number of bits.  To go to a non-power-of-2 basis, things are slightly more complex (but still not bad), the shifts would be replaced by / and the &'s by %.  Code-wise is isn't a lot but it could impose a performance hit.

More difficult would be compacting the data structure on delete.  I don't believe that this is required because we don't delete from things this is likely to replace.